### ISSUE TITLE
Issue on page /api/_generated/jax/netket.jax.expect.html

### ISSUE THREAD

**Issue by Rose-max111** [netket/netket#2039]
Hi, I am trying to rewrite netket.jax.expect in julia for a better performance. And I am quite confuse on how to deal with the unnormalized p.d.f in AD?

As the doc written, we have to calculate the derivative $\nabla \log(p(\sigma))$ and the p.d.f $p(\sigma)$ should be normalized. However, I think the code just calculates the unnormalized log-p.d.f. and throw it to AD. Won't this lead to some errors (the normalization coefficient also depends on parameters)? 

Thank you guys for spending time on my question! The code for backward function is here:

```python
def _expect_bwd(n_chains, chunk_size, in_axes, log_pdf, expected_fun, residuals, dout):
    pars, σ, cost_args, ΔL_σ = residuals
    dL̄, dL̄_stats = dout

    if chunk_size is None:

        def f(pars, σ, *cost_args):
            log_p = log_pdf(pars, σ)
            term1 = jax.vmap(jnp.multiply)(ΔL_σ, log_p)
            term2 = expected_fun(pars, σ, *cost_args)
            out = mpi_mean(term1 + term2, axis=0)
            out = out.sum()
            return out

    else:
        if in_axes is None:
            in_axes = (
                None,
                0,
            ) + tuple(None for _ in cost_args)

        def chunked_f(ΔL_σ, pars, σ, *cost_args):
            log_p = apply_chunked(log_pdf, chunk_size=chunk_size, in_axes=(None, 0))(
                pars, σ
            )
            term1 = jax.vmap(jnp.multiply)(ΔL_σ, log_p)
            term2 = apply_chunked(expected_fun, chunk_size=chunk_size, in_axes=in_axes)(
                pars, σ, *cost_args
            )
            out = mpi_mean(term1 + term2, axis=0)
            out = out.sum()
            return out

        # capture ΔL_σ to not differentiate through it
        f = partial(chunked_f, ΔL_σ)

    _, pb = nkvjp(f, pars, σ, *cost_args)
    grad_f = pb(dL̄)
    return grad_f
```

**Comment by gcarleo** [comment]
the normalization cannot be computed explicitly so you cannot just AD it, you need to express the gradient in terms of expected values over P and gradients of the unnornalized P. Let us know if you get a speedup in Julia....

**Comment by Rose-max111** [comment]
Thank you! I know what I should do now.
